File: backend/agent/tools/move.py
from __future__ import annotations

import os
import shutil
from typing import Any

from agent.tools.base import Tool, _in_cwd


class FileMoveTool(Tool):
    """Move or rename a file or directory. Refuses to cross workspace boundaries."""

    name = "file_move"
    description = (
        "Move or rename a file or directory. Both source and destination must be inside the current directory. "
        "Overwrites the destination if it already exists (use with care)."
    )
    params = {
        "src": {"type": "string", "description": "Source path"},
        "dest": {"type": "string", "description": "Destination path"},
    }

    def is_dangerous(self, args: dict[str, Any]) -> bool:
        return not (_in_cwd(args.get("src", "")) and _in_cwd(args.get("dest", "")))

    async def execute(self, args: dict[str, Any]) -> str:
        src = args.get("src", "")
        dest = args.get("dest", "")
        if not src or not dest:
            return "error: src and dest are required"
        if not _in_cwd(src) or not _in_cwd(dest):
            return "error: path is outside current directory"
        abs_src = os.path.abspath(src)
        abs_dest = os.path.abspath(dest)
        if not os.path.exists(abs_src):
            return f"error: source not found: {src}"
        # Refuse moves that would land inside themselves (mv a a/b is undefined).
        if os.path.commonpath([abs_src, abs_dest]) == abs_src and abs_src != abs_dest:
            return f"error: cannot move {src} into itself"
        try:
            d = os.path.dirname(abs_dest)
            if d:
                os.makedirs(d, exist_ok=True)
            shutil.move(abs_src, abs_dest)
            return f"moved: {src} → {dest}"
        except Exception as e:
            return f"error: {e}"


class FileCopyTool(Tool):
    """Copy a file. Creates intermediate directories. Refuses paths outside the workspace."""

    name = "file_copy"
    description = (
        "Copy a file. Both source and destination must be inside the current directory. "
        "Creates intermediate directories at the destination if needed."
    )
    params = {
        "src": {"type": "string", "description": "Source path"},
        "dest": {"type": "string", "description": "Destination path"},
    }

    def is_dangerous(self, args: dict[str, Any]) -> bool:
        return not (_in_cwd(args.get("src", "")) and _in_cwd(args.get("dest", "")))

    async def execute(self, args: dict[str, Any]) -> str:
        src = args.get("src", "")
        dest = args.get("dest", "")
        if not src or not dest:
            return "error: src and dest are required"
        if not _in_cwd(src) or not _in_cwd(dest):
            return "error: path is outside current directory"
        abs_src = os.path.abspath(src)
        abs_dest = os.path.abspath(dest)
        if not os.path.exists(abs_src):
            return f"error: source not found: {src}"
        if os.path.isdir(abs_src):
            return f"error: source is a directory (only file copy is supported): {src}"
        try:
            d = os.path.dirname(abs_dest)
            if d:
                os.makedirs(d, exist_ok=True)
            shutil.copy2(abs_src, abs_dest)
            return f"copied: {src} → {dest}"
        except Exception as e:
            return f"error: {e}"

File: agent/tools/move.go
package tools

import (
	"io"
	"os"
	"path/filepath"
	"strings"
)

type FileMoveTool struct{}

func (t *FileMoveTool) Name() string { return "file_move" }

func (t *FileMoveTool) Execute(args map[string]any) Result {
	src, _ := args["src"].(string)
	dest, _ := args["dest"].(string)
	if src == "" || dest == "" {
		return R("error: src and dest are required")
	}
	if !inCwd(src) || !inCwd(dest) {
		return R("error: path is outside current directory")
	}
	absSrc, err := filepath.Abs(src)
	if err != nil {
		return R("error: " + err.Error())
	}
	absDest, err := filepath.Abs(dest)
	if err != nil {
		return R("error: " + err.Error())
	}
	if _, err := os.Stat(absSrc); os.IsNotExist(err) {
		return R("error: source not found: " + src)
	}
	if absSrc != absDest && strings.HasPrefix(absDest, absSrc+string(filepath.Separator)) {
		return R("error: cannot move " + src + " into itself")
	}
	if d := filepath.Dir(absDest); d != "" {
		os.MkdirAll(d, 0755)
	}
	if err := os.Rename(absSrc, absDest); err != nil {
		if copyErr := copyMove(absSrc, absDest); copyErr != nil {
			return R("error: " + copyErr.Error())
		}
		os.RemoveAll(absSrc)
	}
	return R("moved: " + src + " → " + dest)
}

type FileCopyTool struct{}

func (t *FileCopyTool) Name() string { return "file_copy" }

func (t *FileCopyTool) Execute(args map[string]any) Result {
	src, _ := args["src"].(string)
	dest, _ := args["dest"].(string)
	if src == "" || dest == "" {
		return R("error: src and dest are required")
	}
	if !inCwd(src) || !inCwd(dest) {
		return R("error: path is outside current directory")
	}
	absSrc, err := filepath.Abs(src)
	if err != nil {
		return R("error: " + err.Error())
	}
	absDest, err := filepath.Abs(dest)
	if err != nil {
		return R("error: " + err.Error())
	}
	fi, err := os.Stat(absSrc)
	if os.IsNotExist(err) {
		return R("error: source not found: " + src)
	}
	if fi.IsDir() {
		return R("error: source is a directory (only file copy is supported): " + src)
	}
	if d := filepath.Dir(absDest); d != "" {
		os.MkdirAll(d, 0755)
	}
	if err := copyFile(absSrc, absDest); err != nil {
		return R("error: " + err.Error())
	}
	return R("copied: " + src + " → " + dest)
}

func copyFile(src, dst string) error {
	s, err := os.Open(src)
	if err != nil {
		return err
	}
	defer s.Close()
	d, err := os.Create(dst)
	if err != nil {
		return err
	}
	defer d.Close()
	if _, err := io.Copy(d, s); err != nil {
		return err
	}
	return d.Close()
}

func copyMove(src, dst string) error {
	s, err := os.Open(src)
	if err != nil {
		return err
	}
	defer s.Close()
	d, err := os.Create(dst)
	if err != nil {
		return err
	}
	defer d.Close()
	_, err = io.Copy(d, s)
	return err
}
